File: memento/src/main/java/com/alexstyl/specialdates/contact/Names.java
package com.alexstyl.specialdates.contact;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class Names implements Iterable<String> {

    private static final String SPACE_OR_SEMICOLUMN = " |-";

    private final List<String> names;

    public static Names from(String names) {
        String[] allNames = names.split(SPACE_OR_SEMICOLUMN);
        return new Names(Arrays.asList(allNames));
    }

    private Names(List<String> names) {
        this.names = names;
    }

    public String getPrimary() {
        return names.get(0);
    }

    public int getCount() {
        return names.size();
    }

    public String get(int index) {
        return names.get(index);
    }

    @Override
    public Iterator<String> iterator() {
        return names.iterator();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Names strings = (Names) o;

        return names.equals(strings.names);

    }

    @Override
    public int hashCode() {
        return names.hashCode();
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/contact/Names.kt
package com.alexstyl.specialdates.contact


data class Names(private val names: List<String>) : Iterable<String> {

    val primary: String
        get() = names[0]

    val count: Int
        get() = names.size

    override fun iterator(): Iterator<String> = names.iterator()

    operator fun get(index: Int): String = names[index]

    companion object {

        private val REGEX = "[@a-zA-Zά-ώΆ-Ώα-ωΑ-Ω\\d]+".toRegex()

        private val cache = mutableMapOf<String, Names>()

        fun parse(input: String): Names {
            if (cache.containsKey(input)) {
                return cache[input]!!
            } else {
                val newNames = Names(REGEX.findAll(input)
                        .map { it.value }
                        .toList())
                cache.put(input, newNames)
                return newNames
            }
        }

    }
}
